File: frontend/src/pages/AdminImportAnalytics.js
import { useEffect, useState } from "react";
import {
  Mail, Users, School, UserCheck, Clock, TrendingUp,
  ArrowRight, AlertTriangle, ChevronDown, ChevronRight, Loader2
} from "lucide-react";
import api from "../lib/api";

function StatCard({ icon: Icon, label, value, sub, color }) {
  return (
    <div
      className="rounded-xl border p-4"
      style={{ backgroundColor: "var(--t-surface)", borderColor: "var(--t-border)" }}
      data-testid={`import-stat-${label.toLowerCase().replace(/\s/g, "-")}`}
    >
      <div className="flex items-center gap-3">
        <div className={`w-10 h-10 rounded-lg flex items-center justify-center ${color}`}>
          <Icon className="w-5 h-5" strokeWidth={1.5} />
        </div>
        <div>
          <p className="text-2xl font-bold" style={{ color: "var(--t-text)" }}>{value}</p>
          <p className="text-xs" style={{ color: "var(--t-text-muted)" }}>{label}</p>
          {sub && <p className="text-[10px] mt-0.5" style={{ color: "var(--t-text-faint)" }}>{sub}</p>}
        </div>
      </div>
    </div>
  );
}

function FunnelStep({ label, value, total, isLast }) {
  const pct = total > 0 ? Math.round((value / total) * 100) : 0;
  const barWidth = total > 0 ? Math.max((value / total) * 100, 4) : 4;
  return (
    <div className="flex items-center gap-3">
      <div className="flex-1">
        <div className="flex items-center justify-between mb-1">
          <span className="text-xs font-medium" style={{ color: "var(--t-text-secondary)" }}>{label}</span>
          <span className="text-xs font-bold" style={{ color: "var(--t-text)" }}>
            {value.toLocaleString()} {total > 0 && value !== total && <span style={{ color: "var(--t-text-faint)" }}>({pct}%)</span>}
          </span>
        </div>
        <div className="h-2.5 rounded-full overflow-hidden" style={{ backgroundColor: "var(--t-border)" }}>
          <div
            className="h-full rounded-full transition-all duration-500"
            style={{ width: `${barWidth}%`, background: "linear-gradient(90deg, #1a8a80, #16a34a)" }}
          />
        </div>
      </div>
      {!isLast && <ArrowRight className="w-4 h-4 flex-shrink-0" style={{ color: "var(--t-text-faint)" }} />}
    </div>
  );
}

function BehaviorStat({ label, value, total, format = "count" }) {
  const display = format === "pct" ? `${value}%` : value.toLocaleString();
  return (
    <div className="flex items-center justify-between py-2 border-b" style={{ borderColor: "rgba(255,255,255,0.05)" }}>
      <span className="text-xs" style={{ color: "var(--t-text-secondary)" }}>{label}</span>
      <span className="text-sm font-semibold" style={{ color: "var(--t-text)" }}>{display}</span>
    </div>
  );
}

function RunRow({ run }) {
  const [expanded, setExpanded] = useState(false);
  const confirmedCount = run.confirmed_school_ids?.length || 0;
  const scanAnalytics = run.scan_analytics || {};
  const confirmAnalytics = run.confirm_analytics || {};

  const statusColor = {
    ready: "#d97706",
    scanning: "#1a8a80",
    aggregating: "#1a8a80",
    failed: "#ef4444",
  };
  const isConfirmed = !!run.confirmed_at;

  return (
    <div
      className="rounded-xl border overflow-hidden"
      style={{ backgroundColor: "var(--t-surface)", borderColor: "var(--t-border)" }}
      data-testid={`import-run-${run.run_id}`}
    >
      <button
        onClick={() => setExpanded(!expanded)}
        className="w-full flex items-center gap-3 p-3 text-left hover:bg-white/[0.02] transition-all"
      >
        {expanded ? <ChevronDown className="w-4 h-4 flex-shrink-0" style={{ color: "var(--t-text-faint)" }} /> :
          <ChevronRight className="w-4 h-4 flex-shrink-0" style={{ color: "var(--t-text-faint)" }} />}
        <div className="flex-1 min-w-0 flex items-center gap-3 flex-wrap">
          <span className="text-sm font-medium truncate" style={{ color: "var(--t-text)" }}>
            {run.user_name || run.user_email || "Unknown"}
          </span>
          <span
            className="text-[10px] font-medium px-2 py-0.5 rounded-full"
            style={{
              backgroundColor: isConfirmed ? "rgba(22,163,74,0.1)" : `${statusColor[run.status] || "#6b7280"}20`,
              color: isConfirmed ? "#16a34a" : (statusColor[run.status] || "#6b7280"),
            }}
          >
            {isConfirmed ? `Imported ${confirmedCount}` : run.status}
          </span>
          {run.error && (
            <span className="text-[10px] font-medium px-2 py-0.5 rounded-full" style={{ backgroundColor: "rgba(239,68,68,0.1)", color: "#ef4444" }}>
              Error
            </span>
          )}
        </div>
        <div className="flex items-center gap-4 text-[11px] flex-shrink-0" style={{ color: "var(--t-text-muted)" }}>
          <span>{run.schools_found || 0} found</span>
          <span>{run.messages_scanned || 0} msgs</span>
          <span>{run.started_at ? new Date(run.started_at).toLocaleDateString("en-US", { month: "short", day: "numeric" }) : "—"}</span>
        </div>
      </button>

      {expanded && (
        <div className="px-4 pb-4 pt-1 border-t space-y-3" style={{ borderColor: "rgba(255,255,255,0.05)" }}>
          {/* Scan details */}
          {scanAnalytics.scan_duration_s != null && (
            <div>
              <p className="text-[10px] uppercase tracking-wider font-bold mb-2" style={{ color: "var(--t-text-faint)" }}>Scan</p>
              <div className="grid grid-cols-2 sm:grid-cols-4 gap-2">
                <MiniStat label="Duration" value={`${scanAnalytics.scan_duration_s}s`} />
                <MiniStat label="Passed Guardrails" value={scanAnalytics.passed_guardrails} />
                <MiniStat label="Filtered" value={scanAnalytics.filtered_by_guardrails} />
                <MiniStat label="Unmapped" value={scanAnalytics.unmapped_count} />
              </div>
            </div>
          )}
          {/* Confirm details */}
          {confirmAnalytics.created_count != null && (
            <div>
              <p className="text-[10px] uppercase tracking-wider font-bold mb-2" style={{ color: "var(--t-text-faint)" }}>Confirm</p>
              <div className="grid grid-cols-2 sm:grid-cols-4 gap-2">
                <MiniStat label="Created" value={confirmAnalytics.created_count} />
                <MiniStat label="Skipped" value={confirmAnalytics.skipped_count} />
                <MiniStat label="Conversion" value={`${confirmAnalytics.conversion_rate}%`} />
                <MiniStat label="Coaches" value={confirmAnalytics.total_coaches_created} />
              </div>
              {confirmAnalytics.skip_reasons && Object.values(confirmAnalytics.skip_reasons).some(v => v > 0) && (
                <div className="mt-2 flex gap-2 flex-wrap">
                  {Object.entries(confirmAnalytics.skip_reasons).filter(([, v]) => v > 0).map(([k, v]) => (
                    <span key={k} className="text-[10px] px-2 py-0.5 rounded-full" style={{ backgroundColor: "rgba(239,68,68,0.08)", color: "#f87171" }}>
                      {k.replace(/_/g, " ")}: {v}
                    </span>
                  ))}
                </div>
              )}
            </div>
          )}
          {/* Confirmed schools */}
          {run.confirmed_school_ids?.length > 0 && (
            <div>
              <p className="text-[10px] uppercase tracking-wider font-bold mb-1" style={{ color: "var(--t-text-faint)" }}>Schools Imported</p>
              <div className="flex flex-wrap gap-1.5">
                {run.confirmed_school_ids.map(s => (
                  <span key={s} className="text-[10px] px-2 py-0.5 rounded-full" style={{ backgroundColor: "rgba(26,138,128,0.1)", color: "#1a8a80" }}>
                    {s}
                  </span>
                ))}
              </div>
            </div>
          )}
          {/* Unmapped domains */}
          {run.unmapped_domains?.length > 0 && (
            <div>
              <p className="text-[10px] uppercase tracking-wider font-bold mb-1" style={{ color: "var(--t-text-faint)" }}>Unmapped Domains</p>
              <div className="flex flex-wrap gap-1.5">
                {run.unmapped_domains.map(d => (
                  <span key={d.domain} className="text-[10px] px-2 py-0.5 rounded-full" style={{ backgroundColor: "rgba(217,119,6,0.1)", color: "#d97706" }}>
                    {d.domain} ({d.count})
                  </span>
                ))}
              </div>
            </div>
          )}
          {/* Error */}
          {run.error && (
            <div className="flex items-start gap-2 p-2 rounded-lg" style={{ backgroundColor: "rgba(239,68,68,0.06)" }}>
              <AlertTriangle className="w-3.5 h-3.5 flex-shrink-0 mt-0.5 text-red-400" />
              <p className="text-[11px] text-red-400">{typeof run.error === "string" ? run.error : run.error?.message || JSON.stringify(run.error)}</p>
            </div>
          )}
        </div>
      )}
    </div>
  );
}

function MiniStat({ label, value }) {
  return (
    <div className="p-2 rounded-lg" style={{ backgroundColor: "rgba(255,255,255,0.02)" }}>
      <p className="text-[10px] mb-0.5" style={{ color: "var(--t-text-faint)" }}>{label}</p>
      <p className="text-sm font-bold" style={{ color: "var(--t-text)" }}>{value ?? "—"}</p>
    </div>
  );
}

export default function AdminImportAnalytics() {
  const [overview, setOverview] = useState(null);
  const [funnel, setFunnel] = useState(null);
  const [behavior, setBehavior] = useState(null);
  const [runs, setRuns] = useState([]);
  const [runsTotal, setRunsTotal] = useState(0);
  const [stages, setStages] = useState({});
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    Promise.all([
      api.get("/admin/import-analytics/overview"),
      api.get("/admin/import-analytics/funnel"),
      api.get("/admin/import-analytics/behavior"),
      api.get("/admin/import-analytics/recent-runs"),
      api.get("/admin/import-analytics/stage-distribution"),
    ]).then(([ov, fn, bh, rn, st]) => {
      setOverview(ov.data);
      setFunnel(fn.data);
      setBehavior(bh.data);
      setRuns(rn.data.runs);
      setRunsTotal(rn.data.total);
      setStages(st.data.stages || {});
    }).catch(() => {}).finally(() => setLoading(false));
  }, []);

  if (loading) {
    return (
      <div className="flex items-center justify-center py-24">
        <Loader2 className="w-8 h-8 animate-spin" style={{ color: "var(--t-text-muted)" }} />
      </div>
    );
  }

  const ov = overview || {};
  const fn = funnel || {};
  const bh = behavior || {};

  return (
    <div className="space-y-6" data-testid="admin-import-analytics">
      {/* Overview Cards */}
      <div className="grid grid-cols-2 lg:grid-cols-3 xl:grid-cols-6 gap-3">
        <StatCard icon={Mail} label="Total Imports" value={ov.total_completed_imports || 0} color="bg-teal-600/15 text-teal-600" />
        <StatCard icon={Users} label="Unique Users" value={ov.unique_users || 0} color="bg-blue-600/15 text-blue-400" />
        <StatCard icon={School} label="Schools Imported" value={ov.total_schools_imported || 0} sub={`${ov.avg_schools_per_run || 0} avg per run`} color="bg-teal-700/15 text-teal-600" />
        <StatCard icon={UserCheck} label="Coaches Created" value={(ov.total_coaches_from_kb || 0) + (ov.total_coaches_from_gmail || 0)} sub={`${ov.total_coaches_from_kb || 0} KB · ${ov.total_coaches_from_gmail || 0} Gmail`} color="bg-amber-600/15 text-amber-400" />
        <StatCard icon={Clock} label="Avg Scan Time" value={`${ov.avg_scan_duration_s || 0}s`} color="bg-violet-600/15 text-violet-400" />
        <StatCard icon={TrendingUp} label="Avg Conversion" value={`${ov.avg_conversion_rate || 0}%`} color="bg-teal-600/15 text-teal-600" />
      </div>

      {/* Funnel + Behavior */}
      <div className="grid grid-cols-1 lg:grid-cols-2 gap-4">
        {/* Funnel */}
        <div className="rounded-xl border p-5" style={{ backgroundColor: "var(--t-surface)", borderColor: "var(--t-border)" }}>
          <h3 className="text-sm font-semibold mb-4" style={{ color: "var(--t-text)" }}>Import Funnel</h3>
          <div className="space-y-3">
            <FunnelStep label="Messages Scanned" value={fn.messages_scanned || 0} total={fn.messages_scanned || 0} />
            <FunnelStep label="Schools Found" value={fn.schools_found || 0} total={fn.messages_scanned || 1} />
            <FunnelStep label="High Confidence" value={fn.high_confidence || 0} total={fn.schools_found || 1} />
            <FunnelStep label="User Selected" value={fn.user_selected || 0} total={fn.high_confidence || 1} />
            <FunnelStep label="Actually Created" value={fn.actually_created || 0} total={fn.user_selected || 1} isLast />
          </div>
        </div>

        {/* User Behavior */}
        <div className="rounded-xl border p-5" style={{ backgroundColor: "var(--t-surface)", borderColor: "var(--t-border)" }}>
          <h3 className="text-sm font-semibold mb-4" style={{ color: "var(--t-text)" }}>User Behavior</h3>
          <div>
            <BehaviorStat label="Consent Screens Shown" value={bh.consent_shown || 0} />
            <BehaviorStat label="Imports Started" value={bh.started || 0} />
            <BehaviorStat label="Start Rate" value={bh.start_rate || 0} format="pct" />
            <BehaviorStat label="Previews Shown" value={bh.preview_shown || 0} />
            <BehaviorStat label="Abandoned at Preview" value={bh.abandoned || 0} />
            <BehaviorStat label="Abandon Rate" value={bh.abandon_rate || 0} format="pct" />
            <BehaviorStat label="Deselections" value={bh.total_deselections || 0} />
            <BehaviorStat label="Reselections" value={bh.total_reselections || 0} />
            <BehaviorStat label="Add Manually Clicks" value={bh.add_manually_clicks || 0} />
          </div>
        </div>
      </div>

      {/* Stage Distribution */}
      {Object.keys(stages).length > 0 && (
        <div className="rounded-xl border p-5" style={{ backgroundColor: "var(--t-surface)", borderColor: "var(--t-border)" }}>
          <h3 className="text-sm font-semibold mb-3" style={{ color: "var(--t-text)" }}>Stage Distribution (Confirmed)</h3>
          <div className="flex gap-3 flex-wrap">
            {Object.entries(stages).map(([stage, count]) => (
              <div key={stage} className="px-4 py-2.5 rounded-xl" style={{ backgroundColor: "rgba(26,138,128,0.06)" }}>
                <p className="text-lg font-bold" style={{ color: "var(--t-text)" }}>{count}</p>
                <p className="text-[10px] uppercase tracking-wider" style={{ color: "var(--t-text-muted)" }}>{stage.replace(/_/g, " ")}</p>
              </div>
            ))}
          </div>
        </div>
      )}

      {/* Recent Runs */}
      <div>
        <h3 className="text-sm font-semibold mb-3" style={{ color: "var(--t-text)" }}>
          Recent Import Runs {runsTotal > 0 && <span style={{ color: "var(--t-text-faint)" }}>({runsTotal})</span>}
        </h3>
        {runs.length === 0 ? (
          <div className="rounded-xl border p-8 text-center" style={{ backgroundColor: "var(--t-surface)", borderColor: "var(--t-border)" }}>
            <Mail className="w-8 h-8 mx-auto mb-2" style={{ color: "rgba(255,255,255,0.1)" }} />
            <p className="text-sm" style={{ color: "var(--t-text-muted)" }}>No import runs yet</p>
          </div>
        ) : (
          <div className="space-y-2">
            {runs.map(run => (
              <RunRow key={run.run_id} run={run} />
            ))}
          </div>
        )}
      </div>
    </div>
  );
}

File: backend/routes/admin_import_analytics.py
"""
Admin Import Analytics — Dashboard for tracking Gmail History Import
performance, user behavior, and conversion funnel.
"""

from fastapi import APIRouter, Depends
from datetime import datetime, timezone, timedelta
from database import db
from admin_guard import require_admin

router = APIRouter(prefix="/api/admin/import-analytics", dependencies=[Depends(require_admin)])


def _serialize(doc: dict) -> dict:
    doc.pop("_id", None)
    return doc


@router.get("/overview")
async def import_overview():
    """Aggregate overview stats across all import runs."""
    # Total completed imports
    total_runs = await db.import_analytics.count_documents({})

    # Aggregate from import_analytics collection
    pipeline = [
        {"$group": {
            "_id": None,
            "total_schools_imported": {"$sum": "$confirm.created_count"},
            "total_schools_skipped": {"$sum": "$confirm.skipped_count"},
            "total_coaches_from_kb": {"$sum": "$confirm.coaches_created_from_kb"},
            "total_coaches_from_gmail": {"$sum": "$confirm.coaches_created_from_gmail"},
            "total_messages_scanned": {"$sum": "$scan.messages_scanned"},
            "avg_scan_duration": {"$avg": "$scan.scan_duration_s"},
            "avg_conversion_rate": {"$avg": "$confirm.conversion_rate"},
            "avg_schools_per_run": {"$avg": "$confirm.created_count"},
        }}
    ]
    results = await db.import_analytics.aggregate(pipeline).to_list(1)
    agg = results[0] if results else {}
    agg.pop("_id", None)

    # Unique users who completed imports
    unique_users = await db.import_analytics.distinct("user_id")

    # Total in-progress or pending runs
    pending_runs = await db.import_runs.count_documents({"status": {"$in": ["scanning", "aggregating"]}})
    ready_not_confirmed = await db.import_runs.count_documents({"status": "ready", "confirmed_at": None})

    return {
        "total_completed_imports": total_runs,
        "unique_users": len(unique_users),
        "pending_runs": pending_runs,
        "ready_not_confirmed": ready_not_confirmed,
        "total_schools_imported": agg.get("total_schools_imported", 0),
        "total_schools_skipped": agg.get("total_schools_skipped", 0),
        "total_coaches_from_kb": agg.get("total_coaches_from_kb", 0),
        "total_coaches_from_gmail": agg.get("total_coaches_from_gmail", 0),
        "total_messages_scanned": agg.get("total_messages_scanned", 0),
        "avg_scan_duration_s": round(agg.get("avg_scan_duration", 0), 1),
        "avg_conversion_rate": round(agg.get("avg_conversion_rate", 0), 1),
        "avg_schools_per_run": round(agg.get("avg_schools_per_run", 0), 1),
    }


@router.get("/funnel")
async def import_funnel():
    """Aggregate funnel data across all imports."""
    pipeline = [
        {"$group": {
            "_id": None,
            "total_messages_scanned": {"$sum": "$funnel.messages_scanned"},
            "total_schools_found": {"$sum": "$funnel.schools_found"},
            "total_high_confidence": {"$sum": "$funnel.high_confidence"},
            "total_user_selected": {"$sum": "$funnel.user_selected"},
            "total_actually_created": {"$sum": "$funnel.actually_created"},
        }}
    ]
    results = await db.import_analytics.aggregate(pipeline).to_list(1)
    agg = results[0] if results else {}
    agg.pop("_id", None)

    return {
        "messages_scanned": agg.get("total_messages_scanned", 0),
        "schools_found": agg.get("total_schools_found", 0),
        "high_confidence": agg.get("total_high_confidence", 0),
        "user_selected": agg.get("total_user_selected", 0),
        "actually_created": agg.get("total_actually_created", 0),
    }


@router.get("/behavior")
async def import_behavior():
    """Aggregate user behavior events."""
    pipeline = [
        {"$group": {"_id": "$event", "count": {"$sum": 1}}}
    ]
    results = await db.import_events.aggregate(pipeline).to_list(20)
    events = {r["_id"]: r["count"] for r in results}

    consent_shown = events.get("import_consent_shown", 0)
    started = events.get("import_started", 0)
    preview_shown = events.get("import_preview_shown", 0)
    confirmed = events.get("import_confirmed", 0)
    abandoned = events.get("import_abandoned", 0)
    deselected = events.get("import_suggestion_deselected", 0)
    reselected = events.get("import_suggestion_reselected", 0)
    add_manually = events.get("import_add_manually_clicked", 0)

    return {
        "consent_shown": consent_shown,
        "started": started,
        "preview_shown": preview_shown,
        "confirmed": confirmed,
        "abandoned": abandoned,
        "start_rate": round(started / max(consent_shown, 1) * 100, 1),
        "abandon_rate": round(abandoned / max(preview_shown, 1) * 100, 1),
        "confirm_rate": round(confirmed / max(preview_shown, 1) * 100, 1),
        "total_deselections": deselected,
        "total_reselections": reselected,
        "add_manually_clicks": add_manually,
    }


@router.get("/recent-runs")
async def recent_runs(limit: int = 20, skip: int = 0):
    """List recent import runs with analytics."""
    cursor = db.import_runs.find(
        {},
        {"_id": 0, "suggestions": 0}
    ).sort("started_at", -1).skip(skip).limit(limit)

    runs = await cursor.to_list(length=limit)
    total = await db.import_runs.count_documents({})

    # Enrich with user email
    user_ids = list({r.get("user_id") for r in runs if r.get("user_id")})
    user_map = {}
    if user_ids:
        user_cursor = db.users.find(
            {"user_id": {"$in": user_ids}},
            {"_id": 0, "user_id": 1, "email": 1, "name": 1}
        )
        async for u in user_cursor:
            user_map[u["user_id"]] = {"email": u.get("email", ""), "name": u.get("name", "")}

    enriched = []
    for r in runs:
        uid = r.get("user_id", "")
        user_info = user_map.get(uid, {})
        enriched.append({
            "run_id": r.get("run_id"),
            "user_email": user_info.get("email", ""),
            "user_name": user_info.get("name", ""),
            "status": r.get("status"),
            "started_at": r.get("started_at"),
            "completed_at": r.get("completed_at"),
            "confirmed_at": r.get("confirmed_at"),
            "messages_scanned": r.get("messages_scanned", 0),
            "schools_found": r.get("schools_found", 0),
            "schools_high_confidence": r.get("schools_high_confidence", 0),
            "confirmed_school_ids": r.get("confirmed_school_ids", []),
            "scan_analytics": r.get("scan_analytics"),
            "confirm_analytics": r.get("confirm_analytics"),
            "unmapped_domains": r.get("unmapped_domains", []),
            "error": r.get("error"),
        })

    return {"runs": enriched, "total": total}


@router.get("/stage-distribution")
async def stage_distribution():
    """Aggregate stage distribution across all confirmed imports."""
    pipeline = [
        {"$match": {"confirm_analytics.stages_confirmed": {"$exists": True}}},
        {"$project": {"stages": {"$objectToArray": "$confirm_analytics.stages_confirmed"}}},
        {"$unwind": "$stages"},
        {"$group": {"_id": "$stages.k", "count": {"$sum": "$stages.v"}}},
        {"$sort": {"count": -1}},
    ]
    results = await db.import_analytics.aggregate(pipeline).to_list(10)
    return {"stages": {r["_id"]: r["count"] for r in results}}
